engine/crates/composition/src/subgraphs/keys.rs

```diff
@@ -61,6 +65,44 @@ impl Subgraphs {
 
         build_selection_set(&selection_set_ast.items, self)
     }
+
+    pub(crate) fn with_nested_key_fields<F>(&mut self, handler: F)
+    where
+        F: FnOnce(&Subgraphs, &mut NestedKeyFields),
+    {
+        let mut nested_key_fields = std::mem::take(&mut self.keys.nested_key_fields);
+        handler(self, &mut nested_key_fields);
+        self.keys.nested_key_fields = nested_key_fields;
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct NestedKeyFields {
+    /// Fields that are part of a nested key key _on another type/entity_. Example:
+    ///
+    /// ```graphql,ignore
+    /// type Entity @key(fields: "name nested { identifier }") {
+    ///   name: String!
+    /// }
+    ///
+    /// type Nested {
+    ///   identifier: ID!
+    /// }
+    ///
+    /// ```
+    ///
+    /// `Nested.identifier` is a nested key field.
+    fields: HashSet<FieldId>,
+
+    /// Objects that are part of keys that are not defined on the object itself.
```

Comment:
Can an object be defined on the object itself? Just asking, this reads a bit weird.

Reply:
I can add a code example.